packages/build/src/plugins/child/mutations.js

```diff
@@ -0,0 +1,86 @@
+'use strict'
+
+const fromEntries = require('@ungap/from-entries')
+const isPlainObj = require('is-plain-obj')
+
+const { addErrorInfo } = require('../../error/info')
+
+// `netlifyConfig` is currently read-only.
+// We will soon add support for mutating specific properties. This requires a
+// `Proxy` to warn plugin authors when mutating read-only properties
+const preventConfigMutations = function (netlifyConfig) {
+  return preventObjectMutations(netlifyConfig, [])
+}
+
+// A `proxy` is recursively applied to readonly properties in `netlifyConfig`
+const preventObjectMutations = function (value, keys) {
+  if (isMutable(keys)) {
+    return value
+  }
+
+  if (Array.isArray(value)) {
+    const array = value.map((item) => preventObjectMutations(item, [...keys, '*']))
+    return addProxy(array, keys)
+  }
+
+  if (isPlainObj(value)) {
+    const object = fromEntries(
+      Object.entries(value).map(([key, item]) => [key, preventObjectMutations(item, [...keys, key])]),
+    )
+    return addProxy(object, keys)
+  }
+
+  return value
+}
+
+const addProxy = function (value, keys) {
+  // eslint-disable-next-line fp/no-proxy
+  return new Proxy(value, getReadonlyProxyHandlers(keys))
+}
+
+// This is called when a plugin author tries to set a `netlifyConfig` property
+// eslint-disable-next-line max-params
+const validateReadonlyProperty = function (method, keys, proxy, key, ...args) {
+  const keysA = [...keys, key]
+  if (isMutable(keysA)) {
+    return Reflect[method](proxy, key, ...args)
+  }
+
+  const propName = keysA.join('.')
+  throwValidationError(`"netlifyConfig.${propName}" is read-only.`)
```

Comment:
In future PRs, we can add specific error messages when trying to mutate before-normalization properties.
For example, when trying to mutate `netlifyConfig.build.functions`, the plugin author should be shown the correct property to mutate instead.